File: Week_1/credit/credit/credit.c
#include "cs50.h"
#include <stdio.h>

int getLength(long number);
bool checkLuhn(long number, int length);
void printCardType(long number, int length);

int main(void)
{
    long card_number;
    do
    {
        card_number = get_long("Please enter the credit card number: ");
    }
    while (card_number <= 0);

    int length = getLength(card_number);

    if (checkLuhn(card_number, length))
    {
        printCardType(card_number, length);
    }
    else
    {
        printf("INVALID\n");
    }
}

int getLength(long number)
{
    int length = 0;
    while (number != 0)
    {
        number /= 10;
        length++;
    }
    return length;
}

bool checkLuhn(long number, int length)
{
    int sum = 0;
    bool alternate = false;

    for (int i = 0; i < length; i++)
    {
        int digit = number % 10;
        if (alternate)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }
        sum += digit;
        alternate = !alternate;
        number /= 10;
    }
    return (sum % 10) == 0;
}

void printCardType(long number, int length)
{
    long start_digits = number;
    while (start_digits >= 100)
    {
        start_digits /= 10;
    }

    if ((length == 15) && (start_digits == 34 || start_digits == 37))
    {
        printf("AMEX\n");
    }
    else if ((length == 16) && (start_digits >= 51 && start_digits <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 16 || length == 13) && (start_digits / 10 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week_6/Problem_Set/sentimental-credit/credit.py
def get_length(number):
    return len(str(number))


def check_luhn(number):
    sum = 0
    alternate = False
    number_str = str(number)[::-1]

    for digit_str in number_str:
        digit = int(digit_str)

        if alternate:
            digit *= 2
            if digit > 9:
                digit -= 9

        sum += digit
        alternate = not alternate

    return sum % 10 == 0


def print_card_type(number):
    start_digits = int(str(number)[:2])

    if (get_length(number) == 15) and (start_digits in [34, 37]):
        print("AMEX")
    elif (get_length(number) == 16) and (51 <= start_digits <= 55):
        print("MASTERCARD")
    elif (get_length(number) in [13, 16]) and (str(number).startswith("4")):
        print("VISA")
    else:
        print("INVALID")


def main():
    while True:
        try:
            card_number = int(input("Number: "))
            if card_number > 0:
                break
        except ValueError:
            pass

    if check_luhn(card_number):
        print_card_type(card_number)
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
